### DISCUSSION TITLE
How does snapcast work at all?

### DISCUSSION THREAD

**Discussion by timkolloch** [music-assistant/support#2397] [Q&A] [answered]
Hey there!

I am trying to use Music Assistant to control a separate snapcast server but I am completely lost in how this should work in any way?
In Music Assistant I am seeing the snapclients without a problem but I cannot start playback on them.

I would expect the audio to play after I selected the speaker but nothign happens?

I only get this error:
`WARNING (MainThread) [music_assistant.providers.snapcast.ffmpeg] write called while process already done` in the MusicAssistant log and some weird stuff like that:

```
2024-05-31 17:44:44.257 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 968, in async_update_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/snapcast/media_player.py", line 272, in state
    return STREAM_STATUS.get(self._client.group.stream_status)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/group.py", line 54, in stream_status
    return self._server.stream(self.stream).status
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/server.py", line 272, in stream
    return self._streams[stream_identifier]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'MusicAssistant--5041'
2024-05-31 17:44:44.265 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 968, in async_update_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/snapcast/media_player.py", line 144, in state
    return STREAM_STATUS.get(self._group.stream_status)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/group.py", line 54, in stream_status
    return self._server.stream(self.stream).status
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/server.py", line 272, in stream
    return self._streams[stream_identifier]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'MusicAssistant--5142'
2024-05-31 17:44:44.268 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 968, in async_update_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/snapcast/media_player.py", line 272, in state
    return STREAM_STATUS.get(self._client.group.stream_status)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/group.py", line 54, in stream_status
    return self._server.stream(self.stream).status
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/server.py", line 272, in stream
    return self._streams[stream_identifier]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'MusicAssistant--5142'
2024-05-31 17:44:46.334 WARNING (MainThread) [homeassistant.components.androidtv_remote] Disconnected from 55OLED854/12 at 192.168.178.170
2024-05-31 17:44:50.175 WARNING (Thread-11) [pychromecast.socket_client] [55OLED854/12(192.168.178.170):8009] Heartbeat timeout, resetting connection
2024-05-31 17:44:50.469 ERROR (SyncWorker_4) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 968, in async_update_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/snapcast/media_player.py", line 144, in state
    return STREAM_STATUS.get(self._group.stream_status)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/group.py", line 54, in stream_status
    return self._server.stream(self.stream).status
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/server.py", line 272, in stream
    return self._streams[stream_identifier]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'MusicAssistant--4989'
2024-05-31 17:44:50.476 ERROR (SyncWorker_4) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 968, in async_update_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1132, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1067, in __async_calculate_state
    state = self._stringify_state(available)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1015, in _stringify_state
    if (state := self.state) is None:
                 ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/snapcast/media_player.py", line 272, in state
    return STREAM_STATUS.get(self._client.group.stream_status)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/group.py", line 54, in stream_status
    return self._server.stream(self.stream).status
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/snapcast/control/server.py", line 272, in stream
    return self._streams[stream_identifier]
           ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
KeyError: 'MusicAssistant--4989'
```

in my homeassistant log, I am pretty clueles but it seems like a new stream with a new port is opened from time to time?

**Comment by OzGav** [comment]
Have you reviewed the troubleshooting section of the docs?

**Reply by timkolloch** [reply]
Yes, my snapserver is version 0.28 so it should work and I can see in the log that the connection was successful.

When I tried to use the inbuilt snapcast server it worked once but I was not able to change the playback to another artist (it this because pausing is not supported and changing the artist is an implicit pause?).

When using the separate snapcast server I sometimes also see that an "External source" is playing on that device with the name of another stream the snapcast server of mine is configured to use. I cannot override that but can also not remove the stream from the device using snapdroid or anything thus I cannot start playing on my device.

Edit: 
I also see `INFO (MainThread) [music_assistant.streams] Start Queue Flow stream for Queue SM-T970 - crossfade: False` when clicking play while my device is selected

Edit 2:

This is the log of my snapcast server, when I try to start playback from MusicAssistant:

```
2024-06-01 10-27-44.945 [Info] (Server) Stream.AddStream("tcp://0.0.0.0:5122?name=MusicAssistant--5122&sampleformat=48000:16:2")
2024-06-01 10-27-44.946 [Info] (PcmStream) PcmStream: MusicAssistant--5122, sampleFormat: 48000:16:2
2024-06-01 10-27-44.946 [Info] (TcpStream) TcpStream host: 0.0.0.0, port: 5122, is server: 1
2024-06-01 10-27-44.946 [Info] (FlacEnc) Init - compression level: 2
2024-06-01 10-27-56.605 [Info] (Server) Stream.RemoveStream("MusicAssistant--5122")
2024-06-01 10-27-56.606 [Error] (TcpStream) Accept failed: Operation canceled
```

Could it be a problem that my Homeassistant is running in a VM and the snapcast server in a Docker container?

**Reply by OzGav** [reply]
> When I tried to use the inbuilt snapcast server it worked once but I was not able to change the playback to another artist (it this because pausing is not supported and changing the artist is an implicit pause?).

I don't know you mean here. MA plays back a queue of tracks. You can add or remove tracks to the queue. Do you mean click on another track in the queue? If so I am doing that now and it works but I do note that sometimes I hear no sound and have to stop and start the stream in the snap client browser window.  

> When using the separate snapcast server I sometimes also see that an "External source" is playing on that device with the name of another stream the snapcast server of mine is configured to use. I cannot override that but can also not remove the stream from the device using snapdroid or anything thus I cannot start playing on my device.

I can't follow what you are trying to describe here

> I also see INFO (MainThread) [music_assistant.streams] Start Queue Flow stream for Queue SM-T970 - crossfade: False when clicking play while my device is selected

This is normal

> Could it be a problem that my Homeassistant is running in a VM and the snapcast server in a Docker container?

If you can just try with the internal snapcast server that will eliminate some of the variables. Most of the issues these days are due to misconfigurations. I run HAOS in Proxmox so that should not be a problem in itself. How you have setup the docker container for your snapserver could be. Why have you done that when MA ships wih its own?

**Reply by timkolloch** [reply]
I was tired when I wrote that and reading it again now it is hard for me to decipher what I wanted to say, I am sorry and will rephrase it with more information.

> I don't know you mean here. MA plays back a queue of tracks. You can add or remove tracks to the queue. Do you mean click on another track in the queue? If so I am doing that now and it works but I do note that sometimes I hear no sound and have to stop and start the stream in the snap client browser window.

I used the inbuilt snapcast server and connected my snapdroid to it. When I started playback of Artist A it worked directly and I was able to listen to the music using my Android client. When I then tried to start playback from Artist B while music of Artist A was still playing nothing happened. Playback of Artist A stopped but no music from Artist B was played. 

This seems to be solved now that I am trying it again probably I was just too unpatient, restarting my Homeassistant and trying again worked even though it takes a few seconds for the playback of Artist B to start which is completely acceptable.

> I can't follow what you are trying to describe here

I tried using my snapcast server running in a docker container instead of the inbuilt one. When I selected my snapdroid client that was connected to the snapcast server I saw this:
![image](https://github.com/music-assistant/hass-music-assistant/assets/46758556/cc7faf73-eb84-4570-b3e3-14b54c718daa)

because the stream "Snapserver" was selected on my device but I was not able to deselect the stream using snapdroid (not a problem of MusicAssistant in any way I guess) and was also not able to "override" the current stream using Music Assistant (but I do not know if this should be possible).

> This is normal

I was expecting that. But together with this output I also get this from the addon:
```
2024-06-03 20:46:05.901 INFO (MainThread) [music_assistant.streams] Start Queue Flow stream for Queue SM-T970 - crossfade: False
2024-06-03 20:46:12.791 WARNING (MainThread) [asyncio] pipe closed by peer or os.write(pipe, data) raised exception.
2024-06-03 20:46:12.794 WARNING (MainThread) [asyncio] pipe closed by peer or os.write(pipe, data) raised exception.
2024-06-03 20:46:12.799 WARNING (MainThread) [asyncio] pipe closed by peer or os.write(pipe, data) raised exception.
2024-06-03 20:46:12.803 WARNING (MainThread) [music_assistant.providers.snapcast.ffmpeg] write called while process already done
```

And this from my snapcast server docker container:
```
2024-06-03 18-57-35.540 [Error] (Server) Server::onMessageReceived JsonRequestException: {"error":{"code":-32603,"data":"Stream not found","message":"Internal error"},"id":288,"jsonrpc":"2.0"}, message: {"id":288,"jsonrpc":"2.0","method":"Group.SetStream","params":{"id":"8d369d8c-8029-2af0-8b21-5fa04cebcdfd","stream_id":"default"}}
2024-06-03 18-57-35.541 [Info] (Server) Stream.AddStream("tcp://0.0.0.0:5023?name=MusicAssistant--5023&sampleformat=48000:16:2")
2024-06-03 18-57-35.542 [Info] (PcmStream) PcmStream: MusicAssistant--5023, sampleFormat: 48000:16:2
2024-06-03 18-57-35.542 [Info] (TcpStream) TcpStream host: 0.0.0.0, port: 5023, is server: 1
2024-06-03 18-57-35.542 [Info] (FlacEnc) Init - compression level: 2
2024-06-03 18-57-47.338 [Info] (Server) Stream.RemoveStream("MusicAssistant--5023")
2024-06-03 18-57-47.339 [Error] (TcpStream) Accept failed: Operation canceled
```

Both logs together give me two thoughts about what might be the problem
1. Some networking error not allowing the stream to be created correctly even though general network connection has to be there otherwise the snapcast server wouldn't know that a stream should be added
2. The log of my snapcast server says "Stream not found" before a Stream should be added, I do not know if that might be a timing issue or something like that? But that would not explain the failure of adding a new stream. 

I hope it is a bit clearer now what I wanted to say and again, sorry for the inconvenience :(

**Comment by SantiagoSotoC** [comment]
Using an external server is an advanced use, it is for some users who need to touch the configuration file or snapcast is the center of their audio system and not Music Assistant. Does your network have something like vlans or is it flat ?

**Reply by timkolloch** [reply]
I used snapcast for my home before when I found Music Assistant so I wanted to do minimal changes to the setup.

My network itself is flat. The only thing is that the snapserver is running in a Docker container and my HomeAssistant is running as HAOS in a VM. Docker container and Home Assistant run on the same physical device.

**Reply by SantiagoSotoC** [reply]
Do you have the snapcast network in host mode ?

**Reply by SantiagoSotoC** [reply]
For now you are required to have a stream named default in your configuration file.
If you can send me the configuration file of your snapserver

**Reply by timkolloch** [reply]
> Do you have the snapcast network in host mode ?

No I did not. I tried that out right now but I never did that before. As I understood a docker container running in network mode host does not use port mappings but uses the ports of the host directly. So I figured that the server should be accessible by 
`ip-of-my-host:1704 or 1705` but the connection timed out.

> For now you are required to have a stream named default in your configuration file.
If you can send me the configuration file of your snapserver

```
[stream]
source = tcp://0.0.0.0?name=Snapserver
source = librespot:///usr/bin/librespot?name=SpotifyA&username=USERNAME&password=PASSWORD&devicename=SpotifyA
source = librespot:///usr/bin/librespot?name=SpotifyB&username=USERNAME&password=PASSWORD&devicename=SpotifyA

[http]
doc_root = /usr/share/snapserver/snapweb
```

Here is the configuration file I used, would it be enough to change the first source to be named "default"?

**Reply by SantiagoSotoC** [reply]
The host mode is required in Docker because MA creates TCP streams in snapcast in order to play audio. 
Dynamically chosen ports are used. 
The error you have is that ma and snapcast can talk to each other on 1705 which is the control port, but Docker blocks the TCP streams and I can't receive the audio.

Yes, in host mode no mappings are used.

Also you have to change the name of the first stream to default

Because your default stream is TCP ? And not a named Pipe ?

**Reply by timkolloch** [reply]
> The host mode is required in Docker because MA creates TCP streams in snapcast in order to play audio. Dynamically chosen ports are used. The error you have is that ma and snapcast can talk to each other on 1705 which is the control port, but Docker blocks the TCP streams and I can't receive the audio.
> 
> Yes, in host mode no mappings are used.
> 
> Also you have to change the name of the first stream to default

But when I run the snapcast server container in host mode no one can connect to it neither my android client nor Music Assistant. With this problem it will not change anything if I change the name of the first stream or am I understanding something wrong?

> Because your default stream is TCP ? And not a named Pipe ?

I do not understand what you mean by that.

**Reply by SantiagoSotoC** [reply]
Using Docker with snapcast is complicated, there is one more permission that you have to give it now I do not remember, I will look for it.

Yes, changing the stream name is not going to work as long as ma and snapserver can't communicate correctly.

**Reply by SantiagoSotoC** [reply]
Here I found how it's done, this is part of the documentation of a Docker image for snapcast

Depending on your system and the security you require, you can allow binding to privileged ports by adding the NET_BIND_SERVICE Linux capability to the container:

docker run ... --cap-add=NET_BIND_SERVICE ...

If the docker daemon itself is not allowed to bind privileged ports, you may also allow the binding system-wide. Grant the privilege for all(!) processes:

sudo sysctl -w net.ipv4.ip_unprivileged_port_start=0

**Reply by timkolloch** [reply]
> Using Docker with snapcast is complicated, there is one more permission that you have to give it now I do not remember, I will look for it.
> 
> Yes, changing the stream name is not going to work as long as ma and snapserver can't communicate correctly.

Okay, thank you very much, I will continuing googling as well!

**Reply by SantiagoSotoC** [reply]
That's why I stopped using it, I use Docker a lot but for snapcast I got bored, it's easier to run it on the host directly.

**Reply by timkolloch** [reply]
> Here I found how it's done, this is part of the documentation of a Docker image for snapcast
> 
> Depending on your system and the security you require, you can allow binding to privileged ports by adding the NET_BIND_SERVICE Linux capability to the container:
> 
> docker run ... --cap-add=NET_BIND_SERVICE ...
> 
> If the docker daemon itself is not allowed to bind privileged ports, you may also allow the binding system-wide. Grant the privilege for all(!) processes:
> 
> sudo sysctl -w net.ipv4.ip_unprivileged_port_start=0

I tried that but it sadly does not work. I think it is because this capability is only needed if I want to bind to ports <=1024 but as snapcast in my case only uses ports 1704, 1705 and 1780 this shouldn't be a problem

> That's why I stopped using it, I use Docker a lot but for snapcast I got bored, it's easier to run it on the host directly.

I begin to understand that :D

**Reply by SantiagoSotoC** [reply]
Just tomorrow I have access to the machine where I have the old docker compose I do not remember if I did not delete it.

You can try to expose a range of ports from 4953 to 5153. 
the format would be as follows 
 -p 4953-5153:4953-5153
in addition to 1704 and 1705
in case you have snapweb 1780 it is also required.

**Reply by timkolloch** [reply]
> Just tomorrow I have access to the machine where I have the old docker compose I do not remember if I did not delete it.
> 
> You can try to expose a range of ports from 4953 to 5153. the format would be as follows -p 4953-5153:4953-5153 in addition to 1704 and 1705 in case you have snapweb 1780 it is also required.

I looked with netstat and see that a port is bound for 1704,1705 and 1780 so I gues this shouldn't be the problem sadly:

```
tcp        0      0 0.0.0.0:1780            0.0.0.0:*               LISTEN      4132631/snapserver  
tcp        0      0 0.0.0.0:1704            0.0.0.0:*               LISTEN      4132631/snapserver  
tcp        0      0 0.0.0.0:1705            0.0.0.0:*               LISTEN      4132631/snapserver
```

**Reply by SantiagoSotoC** [reply]
ok and your machine does not have a firewall ?

**Reply by timkolloch** [reply]
I do not think that a firewall can be the reason. When I run snapcastserver in a docker container with port mapping (1780:1780) it works so traffic is allowed on port 1780 or can it be different if the traffic is not forwarded?

**Reply by SantiagoSotoC** [reply]
So snapweb does not work for you either ?

**Reply by timkolloch** [reply]
> So snapweb does not work for you either ?

When I start it a normal docker network with port mappings `1704:1704, 1705:1705, 1780:1780` everything works as expected. 

When I start the container in `network_mode: host` it does not work. 

When I start the snapserver directly on my host so without docker at all it still does not work.

With "it does not work" I mean that nothing works. I cannot connect snapclients to my server, I cannot reach snapweb, nothing.

I really do not get why it works with docker but not with the other two options.

**Reply by SantiagoSotoC** [reply]
docker publish manipulates iptables, that could be a reason, try to open those ports manually in iptables or use nmap from another machine.

**Reply by timkolloch** [reply]
Ah yes, when I start the docker container in `network_mode: host` the ports seem open with `netstat` but running `nmap` from a different machine shows that the ports are closed. 
However, when I run the container without `network_mode: host` then the ports are shown as open using `nmap`.

I opened the ports directly so now `nmap` also shows the ports as open. I can access snapweb and can connect clients to it. The server sees a new control connection. 

However playback still does not work and I get the same error that I got when I ran the container without `network_mode: host` enabled.

Can this be because the random ports for the MusicAssistant streams are still not opened correctly and can thus not be forwarded to the docker container even in host mode?

**Reply by SantiagoSotoC** [reply]
4953-5153 that is the range of ports that ma uses to send audio to snapcast you have to open them as well

**Reply by timkolloch** [reply, marked as answer]
Thank you very, very, very much, it works now!

In the following section I will comprehend the content of this thread to state a condensed solution to my problem in order to mark it as the answer.

## Problem solution
First of all many thanks to @SantiagoSotoC for the extensive help!
I was running the snapcast server in a docker container which had following port bindings: 1704:1704, 1705:1705 and 1780:1780. This allowed a connection from MusicAssistant to my snapcast server because for this the already forwarded port 1705 is used.

However playing  a stream is done by adding a stream to the snapcast server and then selecting it as the source. As the stream is started on a port between 4953 and 5153 these ports also need to be open to the container. Because you do not want to bind every single port to a container in the `docker-compose.yml` it is better to run the container with `network_mode: host` enabled so that the container directly uses the hosts network stack without docker mappings.

If started with port mappings docker's job is it to open the ports. This does not happen when the network mode is set to host. In order to connect to the server, add streams and play music via them you have to open the ports manually either using `iptables` or `ufw` or some other alternative.

### TL;DR
So for everything to work to the following
1. Run your docker container with `network_mode: host`
2. Open ports 1704, 1705, 1780 using `iptables`, `ufw` or any other alternative so general connection works
3. Also open ports 4953 to 5153 so that MusicAssistant can successfully add the stream to snapcast server and play music via it.

Note: 
Please make sure that you have a stream called "default" in your `snapserver.conf` so that MusicAssistant can switch back to it to stop playback.

**Reply by SantiagoSotoC** [reply]
Glad to help, just one comment, for the moment you have to leave a stream with the default name

**Reply by timkolloch** [reply]
> Glad to help, just one comment, for the moment you have to leave a stream with the default name 

I will update the answer, thanks for the hint!

**Comment by wyllys66** [comment]
Question - 
   I enabled the Snapcast player provider in Music Assistant and now I want to add a snapcast group player. When I select "Snapcast", it appears in the list of players, but with the hourglass icon and says "This player is not (yet) available".
   
 What do I need to do to enable it or configure it?  I am trying to use snapcast as client running on the same RPI as the HA/MA server so I can play audio out of the pi itself through the default HDMI output port and into an amplifier feeding some in-house speakers.

**Reply by OzGav** [reply]
I don't use the client but I do use the web interface. For that I have to click a button to connect to the server. Is the client connected to the server?

Also see here https://music-assistant.io/faq/stream-to/

**Comment by wyllys66** [comment]
I cant seem to figure this out.  It seems like it should be straightforward but its obviously not.

1. I have an RPI5 running the latest HomeAssistant image with Music Assistant installed
2. I added the Snapcast Player Provider that runs the snapcast server on the localhost port 1705

This is where I am struggling:
   I want to add a snapcast client on the same server and have it pipe audio to the local audio output.   I cannot figure out how to do this or if it is even possible.

**Reply by OzGav** [reply]
Yeah sorry can’t help with the Snapcast client as I haven’t heard of anyone doing that. Why can’t you use squeezelite?
